### PR TITLE
Adding an anti-bruteforce mechanism on the administration pages login.

### PR DESCRIPTION
**Pull request by Seb-C**
There doesn't seem to be a better way to return the page's content with the error without manually building a new Response object.

### CODE REVIEW

File: bootstrap.php [hunk 1]
```diff
@@ -55,3 +55,46 @@
...
+            'where' => array(
+                array('logi_created_at', '>=', $time_from),
+                'logi_driver' => 'nos',
+                'logi_state'  => 'fail',
+                'logi_action' => 'login',
+                'logi_ip'     => \NC::remoteIp(),
+            ),
+        ))->count();
+
+        if ($failures >= \Arr::get($wait_failures_config, 'attempts', 3)) {
```
Comment (Meroje) [inline bootstrap.php]: 3 attempts is way too low

**Comment by Meroje** [comment]
A successful login must reset the attempts counter

**Comment by Meroje** [comment]
Also IPs are cheap, there should be a limit (higher ofc) on the whole user

File: bootstrap.php [hunk 2]
```diff
@@ -55,3 +55,46 @@
...
     ));
 });
+
+\Event::register_function('admin.beforeLogin', function ($params) {
+    $controller = \Arr::get($params, 'controllerInstance');
+    $wait_failures_config = \Config::load('novius_loginhistory::wait_after_admin_login_failures', true);
+    $wait_failures_enabled = \Arr::get($wait_failures_config, 'enabled', true);
+
+    if ($wait_failures_enabled) {
+        $time_from = \Date::forge()
```
Comment (Meroje) [inline bootstrap.php]: You're missing the concept of rate here, 3 failed attempts within 300 seconds will ban me for those same 300 seconds. The wait should start on the last failed attempt

**Comment by shaoshiva** [comment]
@Meroje A limit on the user could easily be exploited by an attacker to ban someone.

**Comment by Meroje** [comment]
Yes. Is it worse than getting an account compromised ?

**Comment by shaoshiva** [comment]
Not worse, not better. IP is sufficient.

File: bootstrap.php (hunk 1 above)
Comment (shaoshiva) [inline bootstrap.php]: Would have set 10 tries at least.

**Comment by jguyomard** [comment]
Would be great to add a config to whitelist some ips.

**Comment by shaoshiva** [comment]
Possible to add a warning after 3 attempts ?

:gb: `You have a limited number of attempts before your account is temporarily disabled.`

:fr: `Il vous reste un nombre limité d'essai avant que votre compte ne soit temporairement désactivé.`

File: bootstrap.php (hunk 1 above)
Comment (Seb-C) [inline bootstrap.php]: Done.

File: config/config.php [hunk 3]
```diff
@@ -22,5 +22,9 @@
         'attempts' => 10,
         // Time to wait (seconds)
         'time_to_wait' => 300,
+        // In case you need to setup a mechanism to whitelist someone
+        'is_whitelisted' => function() {
```
Comment (Meroje) [inline config/config.php]: well now it's odd not to provide `is_throttled` (which would return `$failures >= \Arr::get($wait_failures_config, 'attempts', 3)`)

File: bootstrap.php [hunk 4]
```diff
@@ -55,3 +55,47 @@
...
     ));
 });
+
+\Event::register_function('admin.beforeLogin', function ($params) {
+    $controller = \Arr::get($params, 'controllerInstance');
+    $wait_failures_config = \Arr::get(\Config::load('novius_loginhistory::config', true), 'wait_after_admin_login_failures', []);
+    $wait_failures_enabled = \Arr::get($wait_failures_config, 'enabled', true);
+    $callback_is_whitelisted = \Arr::get($wait_failures_config, 'is_whitelisted');
+
+    if ($wait_failures_enabled && is_callable($callback_is_whitelisted) && !$callback_is_whitelisted()) {
```
Comment (jguyomard) [inline bootstrap.php]: if `$callback_is_whitelisted` is not callable, you have to do this test. I think a better condition is:

```
if ($wait_failures_enabled && (!is_callable($callback_is_whitelisted) || !$callback_is_whitelisted())) {
```

**Comment by Meroje** [comment]
woopsies, we need to keep 5.3 compatibility here

**Pull request merged**